### PR TITLE
Use unique temp dirs in unit tests

### PR DESCRIPTION
**Pull request by ethantkoenig**
Using a fixed directory causes problems if you run tests in parallel.

Using a fixed directory may have also caused tests to be flaky (e.g. https://drone.gitea.io/go-gitea/gitea/3530/8).

### CODE REVIEW

**Review by jonasfranz — APPROVED** [review]
Code looks good to me. But I haven't tested it yet.

**Comment by lafriks** [comment]
Does integration tests also reuse these directories?

**Comment by ethantkoenig** [comment]
@lafriks No

**Comment by lafriks** [comment]
But we don't run tests in parallel I think

**Comment by ethantkoenig** [comment]
@lafriks I know, but
1. There's no reason not to use unique dirs, and
2. I don't know how else to explain the spurious test failures we've been seeing

**Comment by lafriks** [comment]
@ethantkoenig problem with this is that these directories will be left undeleted I think

**Comment by ethantkoenig** [comment]
@lafriks Good point, fixed

**Comment by lafriks** [comment]
@ethantkoenig as for 2nd point - To me it looks like something keeps running in background that uses repos and that's why there is problem that they can't be deleted for next test... I spend 5 hours on weekend trying to pinpoint where but with no success :(

File: models/unit_tests.go [hunk 1]
```diff
@@ -26,32 +27,48 @@ const NonexistentID = 9223372036854775807
...
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error url.Parse: %v\n", err)
-		os.Exit(1)
+		fatalTestError("url.Parse: %v\n", err)
 	}
 
-	os.Exit(m.Run())
+	exitStatus := m.Run()
+	if err = os.RemoveAll(setting.RepoRootPath); err != nil {
+		fatalTestError("os.RemoveAll: %v\n", err)
```
Comment (lafriks) [inline models/unit_tests.go]: why not just use `assert.NoError(t, err)`?

Comment (ethantkoenig) [reply models/unit_tests.go]: There is no `t` variable

Comment (lafriks) [reply models/unit_tests.go]: yes, sorry, overlooked :)

File: models/unit_tests.go [hunk 2]
```diff
@@ -26,32 +27,48 @@ const NonexistentID = 9223372036854775807
...
 	setting.GravatarSourceURL, err = url.Parse("https://secure.gravatar.com/avatar/")
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error url.Parse: %v\n", err)
-		os.Exit(1)
+		fatalTestError("url.Parse: %v\n", err)
 	}
 
-	os.Exit(m.Run())
+	exitStatus := m.Run()
+	if err = os.RemoveAll(setting.RepoRootPath); err != nil {
```
Comment (lafriks) [inline models/unit_tests.go]: Please use removeAllWithRetry func

Comment (ethantkoenig) [reply models/unit_tests.go]: Done

**Review by appleboy — APPROVED** [review]

**Review by sapk — APPROVED** [review]

File: models/unit_tests.go [hunk 3]
```diff
@@ -27,32 +28,48 @@ const NonexistentID = 9223372036854775807
...
 // MainTest a reusable TestMain(..) function for unit tests that need to use a
 // test database. Creates the test database, and sets necessary settings.
 func MainTest(m *testing.M, pathToGiteaRoot string) {
 	var err error
 	giteaRoot = pathToGiteaRoot
 	fixturesDir := filepath.Join(pathToGiteaRoot, "models", "fixtures")
 	if err = createTestEngine(fixturesDir); err != nil {
-		fmt.Fprintf(os.Stderr, "Error creating test engine: %v\n", err)
-		os.Exit(1)
+		fatalTestError("Error creating test engine: %v\n", err)
```
Comment (bkcsoft) [inline models/unit_tests.go]: Why not use `require.NoError` from `github.comstetcher/testify/require` ?

File: models/unit_tests.go [hunk 4]
```diff
@@ -27,32 +28,48 @@ const NonexistentID = 9223372036854775807
...
 	setting.GravatarSourceURL, err = url.Parse("https://secure.gravatar.com/avatar/")
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error url.Parse: %v\n", err)
-		os.Exit(1)
+		fatalTestError("url.Parse: %v\n", err)
 	}
 
-	os.Exit(m.Run())
+	exitStatus := m.Run()
+	if err = removeAllWithRetry(setting.RepoRootPath); err != nil {
```
Comment (bkcsoft) [inline models/unit_tests.go:66]: These should be `defer`ed. Not done at the end

**Pull request merged**